core/src/upgrade/map_addr.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::io::Error as IoError;
+use futures::{future, prelude::*};
+use tokio_io::{AsyncRead, AsyncWrite};
+use upgrade::{ConnectionUpgrade, Endpoint};
+use Multiaddr;
+
+/// Applies a closure on the output of a connection upgrade.
+#[inline]
+pub fn map_with_addr<U, F, I, O>(upgrade: U, map: F) -> MapAddr<U, F>
+    where F: FnOnce(I, &Multiaddr) -> O
```

Comment:
`map` only constraints the type parameters in its `ConnectionUpgrade` impl, so for consistency we should do the same here, no?

Reply:
In practice I was getting inference errors for the `Multiaddr`, so I added the constraint.
I don't know what we're supposed to do in theory.

Reply:
Maybe it is worth revisiting `map` at some point and do the same there, just for consistency. I do not have a strong preference, but if one approach causes us problems I am all for the other one ;-)

Reply:
`map` doesn't have the `Multiaddr` parameter, so no inference error.